lib/Sema/TypeCheckAttr.cpp

```diff
@@ -2093,6 +2093,94 @@ static Type getDynamicComparisonType(ValueDecl *value) {
   return interfaceType->removeArgumentLabels(numArgumentLabels);
 }
 
+static OpaqueTypeArchetypeType *getOpaqueResulType(const AbstractFunctionDecl *func) {
+  auto funTy =
+      (func->getDeclContext()->isTypeContext() ? func->getMethodInterfaceType()
+                                               : func->getInterfaceType())
+          ->getCanonicalType()
+          ->getAs<AnyFunctionType>();
+  auto resultTy = funTy->getResult()->getCanonicalType()->getAs<OpaqueTypeArchetypeType>();
+  return resultTy;
+}
+
+static AnyFunctionType::CanParamArrayRef getParameters(const AbstractFunctionDecl *func) {
+  auto funTy =
+      (func->getDeclContext()->isTypeContext() ? func->getMethodInterfaceType()
+                                               : func->getInterfaceType())
+          ->getCanonicalType()
+          ->getAs<AnyFunctionType>();
+  assert(funTy);
+  return funTy->getParams();
+}
+
+static bool
+matchFunctionWithOpaqueResultType(const AbstractFunctionDecl *replacement,
```

Comment:
Maybe instead this could be a flag for TypeBase::matches()? It really doesn't need to handle parameters and results differently, the only difference is when it finds an opaque type, instead of checking for equality it checks if they have the same constraints